### PR TITLE
Removing 2D array usage in Gemm Kernels

### PR DESCRIPTION
**Pull request by CodeplayBen**
Removes 2D array usage in Gemm kernels as it offers no performance improvements, can be confusing/cause problems (due to the row major nature of C++ 2D arrays), and does not decay nicely to a single pointer like 1 dimensional arrays.

### CODE REVIEW

File: src/operations/blas3/gemm_local.hpp [hunk 1]
```diff
@@ -577,9 +576,11 @@ class Gemm<input_t, output_t, DoubleBuffer, NbcA, NbcB, ClSize, TileType,
    * @param reg_res  2D register array used to store the result C
    */
   template <typename InputPointerType>
-  static SYCL_BLAS_INLINE void compute_block_gemm(
-      InputPointerType B, InputPointerType A, element_t (&reg_a)[item_rows],
-      element_t &reg_b, element_t (&reg_res)[item_rows][item_cols]) noexcept {
+  static SYCL_BLAS_INLINE void compute_block_gemm(InputPointerType B,
+                                                  InputPointerType A,
+                                                  element_t (&reg_a)[item_rows],
```
Comment (mehdi-goli) [inline src/operations/blas3/gemm_local.hpp]: you can just use a pointer here.

File: src/operations/blas3/gemm_no_local.hpp [hunk 2]
```diff
@@ -298,7 +298,7 @@ class Gemm<input_t, output_t, DoubleBuffer, NbcA, NbcB, ClSize, tile_type,
       for (index_t i = 0; i < item_rows; ++i) {
         if (do_check<need_check_boundary>(check_boundary(
                 dim_m_c_start + i * wg_rows, dim_n_c_start + j * wg_cols))) {
-          reg_res[i * item_rows + j] = beta * C[i * wg_rows];
+          reg_res[j * item_rows + i] = beta * C[i * wg_rows];
```
Comment (mehdi-goli) [inline src/operations/blas3/gemm_no_local.hpp]: can you adjust the index everywhere to be either  i + j*xxx or j+ i* xxx

File: src/operations/blas3/gemm_local.hpp (hunk 1 above)
Comment (CodeplayBen) [reply src/operations/blas3/gemm_local.hpp]: Fixed!

File: src/operations/blas3/gemm_no_local.hpp (hunk 2 above)
Comment (CodeplayBen) [reply src/operations/blas3/gemm_no_local.hpp]: Fixed!

**Pull request merged**